balancer/main.go

```diff
@@ -0,0 +1,91 @@
+	balancerinformers "k8s.io/autoscaler/balancer/pkg/client/informers/externalversions"
+	"k8s.io/autoscaler/balancer/pkg/controller"
+	cacheddiscovery "k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/dynamic"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/restmapper"
+	scaleclient "k8s.io/client-go/scale"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var (
+	masterURL                  string
+	kubeconfig                 string
+	balancerReprocessPeriodSec int
+)
+
+const (
+	defaultResyncPeriod = 60 * time.Second
+	mapperResetPeriod   = 30 * time.Second
+)
+
+func init() {
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	flag.IntVar(&balancerReprocessPeriodSec, "reprocess-period-sec", 15, "How often (in second) balancers are processed")
+}
+
+func main() {
+	klog.InitFlags(nil)
+	flag.Parse()
+
+	// set up signals so we handle the first shutdown signal gracefully
+	// TODO: handle sigints
+	stopCh := make(chan struct{})
+
+	cfg, err := clientcmd.BuildConfigFromFlags(masterURL, kubeconfig)
+	if err != nil {
+		klog.Fatalf("Error building kubeconfig: %s", err.Error())
+	}
+	klog.V(1).Infof("Starting Balancer for %v", cfg)
+
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		klog.Fatalf("Error building kubernetes clientset: %s", err.Error())
+	}
+
+	balancerClient, err := balancerclientset.NewForConfig(cfg)
+	if err != nil {
+		klog.Fatalf("Error building Balancer clientset: %s", err.Error())
+	}
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, defaultResyncPeriod)
+	balancerInformerFactory := balancerinformers.NewSharedInformerFactory(balancerClient, defaultResyncPeriod)
+
+	cachedClient := cacheddiscovery.NewMemCacheClient(kubeClient.Discovery())
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(cachedClient)
+
+	// Synchronize mapper until stopCh is closed.
+	go wait.Until(func() {
+		restMapper.Reset()
+	}, mapperResetPeriod, stopCh)
+
+	scaleKindResolver := scaleclient.NewDiscoveryScaleKindResolver(kubeClient.Discovery())
+	scaleClient, err := scaleclient.NewForConfig(cfg, restMapper, dynamic.LegacyAPIPathResolverFunc, scaleKindResolver)
+
+	podInformer := kubeInformerFactory.Core().V1().Pods()
+	core := controller.NewCore(controller.NewScaleClient(context.TODO(), scaleClient, restMapper), podInformer)
+
+	controller := controller.NewController(balancerClient,
+		balancerInformerFactory.Balancer().V1alpha1().Balancers(),
+		kubeClient.CoreV1().Events(""),
+		core,
+		time.Duration(balancerReprocessPeriodSec)*time.Second)
+
+	// Start method is non-blocking and runs all registered informers in a dedicated goroutine.
+	kubeInformerFactory.Start(stopCh)
+	balancerInformerFactory.Start(stopCh)
+
+	controller.Run(3, stopCh)
```

Comment:
What is 3? Introduce a constant to make it more readable?

Reply:
Changed to flag.